vllm/engine/llm_engine.py

```diff
@@ -1994,6 +1998,32 @@ def _build_logits_processors(
             sampling_params.logit_bias = None
             sampling_params.allowed_token_ids = None
 
+        if len(sampling_params.bad_words) > 0:
+            bad_words_ids: List[List[int]] = list()
+            tokenizer = self.get_tokenizer(lora_request)
+
+            for bad_word in sampling_params.bad_words:
+                # To prohibit words both at the beginning
+                # and in the middle of text
+                # (related to add_prefix_space tokenizer parameter)
+                for add_prefix_space in [False, True]:
+                    prefix = " " if add_prefix_space else ""
+                    inputs = {"prompt": prefix + bad_word.lstrip()}
+                    prompt_token_ids = tokenizer.encode(
+                        text=inputs["prompt"], add_special_tokens=False)
+
+                    # If no space at the beginning
+                    # or if prefix space produces a new word token
+                    if (not add_prefix_space) or (
+                            add_prefix_space
+                            and prompt_token_ids[0] != bad_words_ids[-1][0] and
+                            len(prompt_token_ids) == len(bad_words_ids[-1])):
+                        bad_words_ids.append(prompt_token_ids)
+
+            no_bad_words_processor = NoBadWordsLogitsProcessor(
+                bad_words_ids=bad_words_ids)
```

Comment:
Perhaps we can factor out this code? Similar to `get_logits_processors`

Reply:
Ok, I think I get the idea... Tried to unify the code.